# Segmenting and Clustering Neighborhoods in the city of Toronto, Canada

## Assignment Part 1: Create dataframe to match

### Import all the required libraries

Pandas, urlopen and BeautifulSoup to be used

In [1]:
# import pandas library for data analsysis
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# import library to handle requests
import requests

# tranform JSON file into a pandas dataframe
from pandas import json_normalize

# import numpy
import numpy as np

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import urlopen to access internet data
from urllib.request import urlopen

# import Beautiful soup to parse html page
from bs4 import BeautifulSoup

#import folium for generating map
import folium

# import k-means from clustering stage
from sklearn.cluster import KMeans

# confirm that all libraries have been imported
print('Libraries imported!')

Libraries imported!


### Obtain data from specified url

Will use urlopen and read() to get the html codes of the link. The html codes will then be parsed through BeautifulSoup to allow me to access the specific table.

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
# use urlopen to open url and read() to obtain html codes
html = urlopen(url).read()
# use BeautifulSoup to parse html
raw_data = BeautifulSoup(html,'html.parser')

### Obtain column header from web table using the html tags 'th'

The table column names are found to be wrapped within the html tags 'th'. So, I'll go ahead and get it and feed it into a column_header list.

In [3]:
column_tags = raw_data('th')
# create empty list for column header
column_header = []
# iterate across tags, get the text, strip \n at the end and append to column header list
for tags in column_tags:
    header = tags.text.rstrip()
    column_header.append(header)

### Initialize dataframe using the column names

The dataframe is then initialised using the column_header list, dropping 'Canadian postal codes' as it is not required as part of the assignment.

In [4]:
df = pd.DataFrame(columns = column_header).drop(columns='Canadian postal codes')
df

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

### Obtain row data from each row using the html tags 'tr'

Each row of information is found between the html tags 'tr'

In [5]:
row_tags = raw_data('tr')
# iterate across each row tags except 0 as this is the column header row and the last 4 rows which is not relevant
for tags in row_tags[1:len(row_tags)-4]:
        row_postal_code = tags('td')[0].text.rstrip()
        row_borough = tags('td')[1].text.rstrip()
        row_neighborhood = tags('td')[2].text.rstrip()
        #ignore row where borough is not assigned
        if row_borough != 'Not assigned':
            #append each borough assigned row into dataframe
            df = df.append({'Postal Code':row_postal_code,
                            'Borough':row_borough,
                            'Neighborhood':row_neighborhood},ignore_index=True)
df.head(15)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

### Explore how many boroughs and postal code within each borough

In [6]:
print('There are {} boroughs, {} neighborhood and {} postal codes in Toronto.'.
      format(len(df['Borough'].unique()),len(df['Neighborhood'].unique()),len(df['Postal Code'].unique())))

There are 10 boroughs, 98 neighborhood and 103 postal codes in Toronto.


Seems like there are duplicates of neighborhood in a couple of postal codes in Toronto. Let's find out what they are

In [7]:
df.groupby(['Borough','Neighborhood']).count().sort_values(by=['Postal Code'],ascending=False).head()

Postal Code
Borough         Neighborhood                                         
North York      Downsview                                           4
                Don Mills                                           2
                Willowdale                                          2
Central Toronto Davisville                                          1
North York      North Park, Maple Leaf Park, Upwood Park            1

Seems like Donsview, Don Mills and Willowdale has multiple postal codes. To make it distinct we can append the postal codes at the end of the neighborhoods

In [8]:
df['Neighborhood'] = df[['Neighborhood','Postal Code']].apply(lambda x: '-'.join(x),axis=1)

Let's check again

In [9]:
df.groupby(['Borough','Neighborhood']).count().sort_values(by=['Postal Code'],ascending=False).head()

Postal Code
Borough         Neighborhood                                     
Central Toronto Davisville North-M4P                            1
North York      Lawrence Manor, Lawrence Heights-M6A            1
Scarborough     Agincourt-M1S                                   1
North York      York Mills, Silver Hills-M2L                    1
                York Mills West-M2P                             1

Now I have distinct neighborhoods, it will allow me to map out venues in distinctive postal code areas. Another solution is to remove rows of duplicate neighborhoods but I decided to keep it going forwards. 

In [10]:
print(df.shape)

(103, 3)


## Assignment Part II: Combining previous dataframe with geospatial data

I am going to use the geosptial_data.csv instead of Geocoder due to inability to obtain data from the library

### Download csv file and read geospatial data

In [11]:
!wget -q -O Geospatial_data.csv https://cocl.us/Geospatial_data    

### Read geospatial.csv file

In [12]:
geo_spatial = pd.read_csv('Geospatial_data.csv')
geo_spatial.head(10)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

### Merge geo_spatial data with main df

In [13]:
df = pd.merge(df,geo_spatial,how='left',on=['Postal Code'])

In [14]:
print(df.shape)
df.head(10)

(103, 5)


Postal Code           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   

                                      Neighborhood   Latitude  Longitude  
0                                    Parkwoods-M3A  43.753259 -79.329656  
1                             Victoria Village-M4A  43.725882 -79.315572  
2                    Regent Park, Harbourfront-M5A  43.654260 -79.360636  
3             Lawrence Manor, Lawrence Heights-M6A  43.718518 -79.464763  
4  Queen's Park, Ontario Provincial Government-M7A  43.662301 -79.389494  
5                             Islington Avenue-M9A  43.667856 -79.532242  
6                               Malvern, Rouge-M1B  43.806686 -79.194353  
7                                    Don Mills-M3B  43.745906 -79.352188  
8              Parkview Hill, Woodbine Gardens-M4B  43.706397 -79.309937  
9                     Garden District, Ryerson-M5B  43.657162 -79.378937

## Assignment Part III: Exploration and clustering of neighborhoods in Toronto

### Use geopy to obtain longitude and latitude of Toronto city for use in folium

In [15]:
from geopy.geocoders import Nominatim

Define instance for geocoder and name agent at 'Toronto_explorer'

In [16]:
address = 'Toronto City, Canada'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


Create map of Toronto City with neighborhoods superimposed on top. The neighborhoods corresponds to each postal code and note that neighborhoods Donsview, Don Mills and Willowdale has multiple postal codes.

In [17]:
# create map of Toronto City using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

I am going to explore all the neighborhoods and segment them by using the Foursquare API

### Explore one of Toronto's neighborhood

Let's explore one of the neighbourhood(s) in Toronto. We will use the first one. Let's get the latitude and longitude for that.

In [20]:
neighborhood_latitude = df.loc[0,'Latitude']
neighborhood_longitude = df.loc[0,'Longitude']
neighborhood_name = df.loc[0,'Neighborhood']

print('The {} neighborhood(s) has the latitude of {} and longitude of {}'.format(neighborhood_name,
                                                                                neighborhood_latitude,
                                                                                neighborhood_longitude))

The Parkwoods-M3A neighborhood(s) has the latitude of 43.7532586 and longitude of -79.3296565


Let's get the top 100 venues in that area with a 1000 metre radius. Create a GET url and obtain json

In [23]:
radius = 1000
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ec19d40a2e538001b213980'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 29,
  'suggestedBounds': {'ne': {'lat': 43.762258609000014,
    'lng': -79.31721997969855},
   'sw': {'lat': 43.74425859099999, 'lng': -79.34209302030145}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b8991cbf964a520814232e3',
       'name': "Allwyn's Bakery",
       'location': {'address': '81 Underhill drive',
        'lat': 43.75984035203157,
        'lng': -79.32471879917513,
        'labeledLatLngs': [{'label': 'display'

Borrowing the get_category_type function from earlier lab.

In [24]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Clean and restructure json data into a dataframme

In [25]:
venues = results['response']['groups'][0]['items']

nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0    Allwyn's Bakery  Caribbean Restaurant  43.759840 -79.324719
1    Brookbanks Park                  Park  43.751976 -79.332140
2        Tim Hortons                  Café  43.760668 -79.326368
3                A&W  Fast Food Restaurant  43.760643 -79.326865
4  Bruno's valu-mart         Grocery Store  43.746143 -79.324630

Now I have the venues for one neighborhood. How many venues were returned by Foursquare?

In [26]:
print('{} venues were returned by Foursquare for the {} neighborhood.'.format(nearby_venues.shape[0],
                                                                             neighborhood_name))

29 venues were returned by Foursquare for the Parkwoods-M3A neighborhood.


### Explore all of Toronto's neighborhood

Borrowing function from earlier labs to repeat the above steps for all neighborhoods in Toronto

In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
                   
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Use above defined function and run it for all neighborhoods in Toronto and put it to *toronto_venues*

In [28]:
print(df.shape)
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )


(103, 5)
Parkwoods-M3A
Victoria Village-M4A
Regent Park, Harbourfront-M5A
Lawrence Manor, Lawrence Heights-M6A
Queen's Park, Ontario Provincial Government-M7A
Islington Avenue-M9A
Malvern, Rouge-M1B
Don Mills-M3B
Parkview Hill, Woodbine Gardens-M4B
Garden District, Ryerson-M5B
Glencairn-M6B
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale-M9B
Rouge Hill, Port Union, Highland Creek-M1C
Don Mills-M3C
Woodbine Heights-M4C
St. James Town-M5C
Humewood-Cedarvale-M6C
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood-M9C
Guildwood, Morningside, West Hill-M1E
The Beaches-M4E
Berczy Park-M5E
Caledonia-Fairbanks-M6E
Woburn-M1G
Leaside-M4G
Central Bay Street-M5G
Christie-M6G
Cedarbrae-M1H
Hillcrest Village-M2H
Bathurst Manor, Wilson Heights, Downsview North-M3H
Thorncliffe Park-M4H
Richmond, Adelaide, King-M5H
Dufferin, Dovercourt Village-M6H
Scarborough Village-M1J
Fairview, Henry Farm, Oriole-M2J
Northwood Park, York University-M3J
East Toronto-M4J
Harbourfront 

Checking the size of the resulting dataframe

In [29]:
print(toronto_venues.shape)
toronto_venues.head()

(4911, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Parkwoods-M3A              43.753259              -79.329656   
1  Parkwoods-M3A              43.753259              -79.329656   
2  Parkwoods-M3A              43.753259              -79.329656   
3  Parkwoods-M3A              43.753259              -79.329656   
4  Parkwoods-M3A              43.753259              -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3                A&W       43.760643       -79.326865  Fast Food Restaurant  
4  Bruno's valu-mart       43.746143       -79.324630         Grocery Store

Group the venues by neighborhood to check how many venues are found per neighborhood

In [30]:
print(toronto_venues.groupby('Neighborhood').count().shape)
toronto_venues.groupby('Neighborhood').count()

(102, 6)


Neighborhood Latitude  \
Neighborhood                                                                
Agincourt-M1S                                                          44   
Alderwood, Long Branch-M8W                                             27   
Bathurst Manor, Wilson Heights, Downsview North...                     30   
Bayview Village-M2K                                                    14   
Bedford Park, Lawrence Manor East-M5M                                  41   
Berczy Park-M5E                                                       100   
Birch Cliff, Cliffside West-M1N                                        14   
Brockton, Parkdale Village, Exhibition Place-M6K                      100   
Business reply mail Processing Centre-M7Y                              49   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks-M6E                                                24   
Canada Post Gateway Processing Centre-M7R                              55   
Cedarbrae-M1H                                                          30   
Central Bay Street-M5G                                                100   
Christie-M6G                                                          100   
Church and Wellesley-M4Y                                              100   
Clarks Corners, Tam O'Shanter, Sullivan-M1T                            33   
Cliffside, Cliffcrest, Scarborough Village West...                     12   
Commerce Court, Victoria Hotel-M5L                                    100   
Davisville North-M4P                                                  100   
Davisville-M4S                                                        100   
Del Ray, Mount Dennis, Keelsdale and Silverthor...                     17   
Don Mills-M3B                                                          32   
Don Mills-M3C                                                          41   
Dorset Park, Wexford Heights, Scarborough Town ...                     46   
Downsview-M3K                                                          21   
Downsview-M3L                                                           9   
Downsview-M3M                                                           4   
Downsview-M3N                                                          35   
Dufferin, Dovercourt Village-M6H                                       67   
East Toronto-M4J                                                       95   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     18   
Fairview, Henry Farm, Oriole-M2J                                       44   
First Canadian Place, Underground city-M5X                            100   
Forest Hill North & West-M5P                                           46   
Garden District, Ryerson-M5B                                          100   
Glencairn-M6B                                                          34   
Golden Mile, Clairlea, Oakridge-M1L                                    27   
Guildwood, Morningside, West Hill-M1E                                  25   
Harbourfront East, Union Station, Toronto Islan...                    100   
High Park, The Junction South-M6P                                     100   
Hillcrest Village-M2H                                                  20   
Humber Summit-M9L                                                      10   
Humberlea, Emery-M9M                                                    8   
Humewood-Cedarvale-M6C                                                 26   
India Bazaar, The Beaches West-M4L                                     82   
Islington Avenue-M9A                                                   12   
Kennedy Park, Ionview, East Birchmount Park-M1K                        30   
Kensington Market, Chinatown, Grange Park-M5T                         100   
Kingsview Village, St. Phillips, Martin Grove G...                     16   
Lawrence Manor, Lawrence Heights-M6A            

Seems like there is a neighborhood which has not return venues at all. Which one is it?

In [31]:
venues_grouped = toronto_venues.groupby('Neighborhood').count().reset_index()
missing_neigh=df[~df.Neighborhood.isin(venues_grouped.Neighborhood)]
missing_neigh

Postal Code      Borough     Neighborhood   Latitude  Longitude
95         M1X  Scarborough  Upper Rouge-M1X  43.836125 -79.205636

Let's confirm by trying a Foursquare API call

In [32]:
missing_latitude = missing_neigh.loc[95,'Latitude']
missing_longitude = missing_neigh.loc[95,'Longitude']
radius = 1000
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, missing_latitude, missing_longitude, VERSION, radius, LIMIT)
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ec19e2b9388d7001c2709d8'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Rouge',
  'headerFullLocation': 'Rouge, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 0,
  'suggestedBounds': {'ne': {'lat': 43.84512470900002,
    'lng': -79.19318232223782},
   'sw': {'lat': 43.827124690999995, 'lng': -79.21808987776217}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': []}]}}

There was indeed no results. We'll have to ensure that this row of Postal Code M1X and Neighborhood Upper Rouge is not reflected later.

Investigate how many unique categories there are from all returned values

In [33]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 327 uniques categories.


Analyzing each neighborhood through one hot encoding

In [34]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = list(toronto_onehot)
fixed_columns.insert(0,fixed_columns.pop(fixed_columns.index('Neighborhood')))
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0  Parkwoods-M3A                  0                  0        0   
1  Parkwoods-M3A                  0                  0        0   
2  Parkwoods-M3A                  0                  0        0   
3  Parkwoods-M3A                  0                  0        0   
4  Parkwoods-M3A                  0                  0        0   

   Airport Lounge  American Restaurant  Amphitheater  Animal Shelter  \
0               0                    0             0               0   
1               0                    0             0               0   
2               0                    0             0               0   
3               0                    0             0               0   
4               0                    0             0               0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  Auto Garage  \
0                 0                   0                0            0   
1                 0                   0                0            0   
2                 0                   0                0            0   
3                 0                   0                0            0   
4                 0                   0                0            0   

   Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  Badminton Court  \
0              0                0          0           0                0   
1              0                0          0           0                0   
2              0                0          0           0                0   
3              0                0          0           0                0   
4              0                0          0           0                0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Stadium  Beach  Beach Bar  Beer Bar  Beer Store  \
0                   0      0          0         0           0   
1                   0      0          0         0           0   
2                   0      0          0         0           0   
3                   0      0          0         0           0   
4                   0      0          0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Bookstore  Boutique  \
0                   0          0       0           0          0         0   
1                   0          0       0           0          0         0   
2                   0          0       0           0          0         0   
3                   0          0       0           0          0         0   
4                   0          0       0           0          0         0   

   Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0              0                     0               0        0            0   
1              0                     0               0        0            0   
2              0                     0               0        0            0   
3              0                     0               0        0            0   
4              0                     0               0        0            0   

   Bridge  Bubble Tea Shop  Buffet  Burger Joint  Burrito P

What is the shape of the new toronto dataframe?

In [35]:
toronto_onehot.shape

(4911, 327)

The shape matches the total number of venue categories and number of venues found. Now let's group them by the neighborhood to get an idea of the frequency of venue occurences in each category

In [36]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                      Agincourt-M1S                0.0   
1                         Alderwood, Long Branch-M8W                0.0   
2  Bathurst Manor, Wilson Heights, Downsview Nort...                0.0   
3                                Bayview Village-M2K                0.0   
4              Bedford Park, Lawrence Manor East-M5M                0.0   

   Afghan Restaurant  Airport  Airport Lounge  American Restaurant  \
0                0.0      0.0             0.0             0.022727   
1                0.0      0.0             0.0             0.000000   
2                0.0      0.0             0.0             0.000000   
3                0.0      0.0             0.0             0.000000   
4                0.0      0.0             0.0             0.024390   

   Amphitheater  Animal Shelter  Antique Shop  Aquarium  Art Gallery  \
0           0.0             0.0           0.0       0.0          0.0   
1           0.0             0.0           0.0       0.0          0.0   
2           0.0             0.0           0.0       0.0          0.0   
3           0.0             0.0           0.0       0.0          0.0   
4           0.0             0.0           0.0       0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0         0.0                  0.0          0.022727                 0.0   
1         0.0                  0.0          0.000000                 0.0   
2         0.0                  0.0          0.000000                 0.0   
3         0.0                  0.0          0.000000                 0.0   
4         0.0                  0.0          0.000000                 0.0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0              0.0          0.0            0.0              0.0        0.0   
1              0.0          0.0            0.0              0.0        0.0   
2              0.0          0.0            0.0              0.0        0.0   
3              0.0          0.0            0.0              0.0        0.0   
4              0.0          0.0            0.0              0.0        0.0   

   Baby Store  Badminton Court  Bagel Shop    Bakery      Bank  Bar  \
0     0.00000         0.022727     0.00000  0.045455  0.022727  0.0   
1     0.00000         0.000000     0.00000  0.000000  0.000000  0.0   
2     0.00000         0.000000     0.00000  0.000000  0.066667  0.0   
3     0.00000         0.000000     0.00000  0.000000  0.142857  0.0   
4     0.02439         0.000000     0.02439  0.024390  0.048780  0.0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  Beach Bar  \
0             0.0               0.0                 0.0    0.0        0.0   
1             0.0               0.0                 0.0    0.0        0.0   
2             0.0               0.0                 0.0    0.0        0.0   
3             0.0               0.0                 0.0    0.0        0.0   
4             0.0               0.0                 0.0    0.0        0.0   

   Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  Board Shop  \
0       0.0         0.0                 0.0        0.0     0.0         0.0   
1       0.0         0.0                 0.0        0.0     0.0         0.0   
2       0.0         0.0                 0.0        0.0     0.0         0.0   
3       0.0         0.0                 0.0        0.0     0.0         0.0   
4       0.0         0.0                 0.0        0.0     0.0         0.0   

   Bookstore  Boutique  Bowling Alley  Brazilian Restaurant  Breakfast Spot  \
0        0.0       0.0            0.0                   0.0        0.022727   
1        0.0       0.0            0.0                   0.0        0.000000   
2        0.0       0.0            0.0                   0.0        0.000000   
3        0.0       0.0            0.0                   0.0        0.000000   
4        0.0       0.0            0.0                  

Let's check the shape to make sure it is valid

In [37]:
toronto_grouped.shape

(102, 327)

The shape corresponds to the main df shape of 103 minus the neighborhood Upper Rouge which did not return any result from Foursquare APO

Borrowing function to rate the top 10 location for each neighborhood

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now to create a dataframe of neighborhoods with the top 10 most common venue categories

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
# match new dataframe 'Neighborhood' column with the toronto_grouped dataframe
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

(102, 11)


Neighborhood 1st Most Common Venue  \
0                                      Agincourt-M1S    Chinese Restaurant   
1                         Alderwood, Long Branch-M8W        Discount Store   
2  Bathurst Manor, Wilson Heights, Downsview Nort...           Pizza Place   
3                                Bayview Village-M2K   Japanese Restaurant   
4              Bedford Park, Lawrence Manor East-M5M    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery  Caribbean Restaurant   
1              Pharmacy                  Park     Convenience Store   
2                  Park           Coffee Shop                  Bank   
3         Grocery Store           Gas Station                  Bank   
4           Coffee Shop                  Bank                  Park   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Motorcycle Shop                  Bank           Coffee Shop   
1           Pizza Place           Coffee Shop                   Pub   
2              Pharmacy      Community Center   Fried Chicken Joint   
3    Chinese Restaurant            Restaurant         Shopping Mall   
4        Sandwich Place  Fast Food Restaurant            Bagel Shop   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0      Sushi Restaurant           Supermarket  Latin American Restaurant  
1        Sandwich Place                 Trail              Shopping Mall  
2      Sushi Restaurant           Supermarket   Mediterranean Restaurant  
3                  Café                  Park                      Trail  
4                Bakery             Juice Bar           Sushi Restaurant

### Now time to cluster these neighborhoods

Run k-means to cluster the neighborhood into 6 clusters.

In [40]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 4, 4, 4, 5, 5, 4, 5, 5, 5])

Now lets combine the cluster labels with the top 10 venues

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# create a merge dataframe but drop
toronto_merged = df.drop(index=95)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   

                                      Neighborhood   Latitude  Longitude  \
0                                    Parkwoods-M3A  43.753259 -79.329656   
1                             Victoria Village-M4A  43.725882 -79.315572   
2                    Regent Park, Harbourfront-M5A  43.654260 -79.360636   
3             Lawrence Manor, Lawrence Heights-M6A  43.718518 -79.464763   
4  Queen's Park, Ontario Provincial Government-M7A  43.662301 -79.389494   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               4                  Park         Shopping Mall   
1               2           Coffee Shop  Gym / Fitness Center   
2               5           Coffee Shop               Theater   
3               5        Clothing Store  Fast Food Restaurant   
4               5           Coffee Shop      Sushi Restaurant   

  3rd Most Common Venue  4th Most Common Venue  5th Most Common Venue  \
0     Convenience Store               Pharmacy               Bus Stop   
1          Hockey Arena  Portuguese Restaurant                 Lounge   
2                  Café             Restaurant                   Park   
3           Coffee Shop             Restaurant  Vietnamese Restaurant   
4   Japanese Restaurant                   Park        Thai Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Fish & Chips Shop           Supermarket     Food & Drink Shop   
1                  Park            Playground           Pizza Place   
2                 Diner    Italian Restaurant                   Pub   
3          Dessert Shop   Fried Chicken Joint      Sushi Restaurant   
4             Gastropub      Ramen Restaurant                  Café   

    9th Most Common Venue 10th Most Common Venue  
0    Fast Food Restaurant           Skating Rink  
1             Golf Course            Men's Store  
2                  Bakery         Breakfast Spot  
3  Furniture / Home Store            Event Space  
4      Italian Restaurant         Ice Cream Shop

Now let's visualise the clusters

In [42]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining the clusters

__Cluster 1__

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood   1st Most Common Venue  \
6                                  Malvern, Rouge-M1B              Restaurant   
8                 Parkview Hill, Woodbine Gardens-M4B                 Brewery   
10                                      Glencairn-M6B           Grocery Store   
18              Guildwood, Morningside, West Hill-M1E             Pizza Place   
26                                      Cedarbrae-M1H             Coffee Shop   
29                               Thorncliffe Park-M4H             Coffee Shop   
32                            Scarborough Village-M1J          Ice Cream Shop   
34                Northwood Park, York University-M3J             Coffee Shop   
38    Kennedy Park, Ionview, East Birchmount Park-M1K          Discount Store   
44                Golden Mile, Clairlea, Oakridge-M1L            Intersection   
56  Del Ray, Mount Dennis, Keelsdale and Silvertho...  Furniture / Home Store   
60                                      Downsview-M3N                   Hotel   
64                                         Weston-M9N             Pizza Place   
65  Dorset Park, Wexford Heights, Scarborough Town...             Coffee Shop   
70                                      Westmount-M9P             Pizza Place   
71                              Wexford, Maryvale-M1R            Intersection   
77  Kingsview Village, St. Phillips, Martin Grove ...                Pharmacy   
78                                      Agincourt-M1S      Chinese Restaurant   
82        Clarks Corners, Tam O'Shanter, Sullivan-M1T             Coffee Shop   
85  Milliken, Agincourt North, Steeles East, L'Amo...      Chinese Restaurant   
89  South Steeles, Silverstone, Humbergate, Jamest...           Grocery Store   
90                  Steeles West, L'Amoreaux West-M1W      Chinese Restaurant   

        2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
6                       Trail             Coffee Shop  Fast Food Restaurant   
8                 Pizza Place    Fast Food Restaurant  Gym / Fitness Center   
10       Fast Food Restaurant                     Gym           Pizza Place   
18                       Bank    Fast Food Restaurant           Coffee Shop   
26                       Bank             Gas Station     Indian Restaurant   
29          Indian Restaurant           Grocery Store          Burger Joint   
32          Convenience Store              Restaurant         Bowling Alley   
34                Pizza Place  Furniture / Home Store            Restaurant   
38         Chinese Restaurant             Coffee Shop           Pizza Place   
44                     Bakery                Pharmacy           Coffee Shop   
56             Discount Store           Grocery Store    Italian Restaurant   
60              Grocery Store    Fast Food Restaurant           Coffee Shop   
64              Train Station             Coffee Shop   Fried Chicken Joint   
65                 Restaurant  Furniture / Home Store     Electronics Store   
70                Gas Station          Breakfast Spot          Intersection   
71  Middle Eastern Restaurant             Pizza Place         Grocery Store   
77                       Bank            Intersection              Bus Line   
78              Shopping Mall                  Bakery  Caribbean Restaurant   
82                Pizza Place                Pharmacy        Sandwich Place   
85                Pizza Place                    Park            Hobby Shop   
89                Pizza Place                Pharmacy              Gym Pool   
90                Pizza Place             Coffee Shop  Fast Food Restaurant   

         5th Most Common Venue 6th Most Common Venue  \
6                          Spa    Chinese Restaurant   
8   Construction & Landscaping    Athletics & Sports   
10                 Gas Station           Coffee Shop   
18            Greek Restaurant              Bus Line   
26                      Bakery          Intersection   
29                     

Cluster 1 seems to cover areas which generally has everything from coffee shops, restaurants, banks, gyms, parks and etc. Looking at the map, it seems to be mostly outside Toronto City centre hence the assumption is that cluster 1 is mostly to serve __General Residential__ areas.

__Cluster 2__

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
45  York Mills, Silver Hills-M2L                  Park                  Pool   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45                   Zoo        Farmers Market   Dumpling Restaurant   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45  Eastern European Restaurant     Electronics Store  Ethiopian Restaurant   

   9th Most Common Venue 10th Most Common Venue  
45           Event Space     Falafel Restaurant

Cluster 2 covers areas of greatest recreational public amenities like parks, pools and zoo. Cluster 2 can be assumed to be __Recreational__

__Cluster 3__

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
1                                Victoria Village-M4A           Coffee Shop   
14                               Woodbine Heights-M4C                  Park   
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...           Coffee Shop   
22                                         Woburn-M1G                  Park   
40                                      Downsview-M3K           Coffee Shop   
49       North Park, Maple Leaf Park, Upwood Park-M6L           Coffee Shop   
61                                  Lawrence Park-M4N          College Quad   
63                  Runnymede, The Junction North-M6N           Coffee Shop   
66                                York Mills West-M2P                  Park   
91                                       Rosedale-M4W           Coffee Shop   
94                                      Northwest-M9W                 Hotel   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
1   Gym / Fitness Center          Hockey Arena      Portuguese Restaurant   
14           Coffee Shop        Sandwich Place                Pizza Place   
17             Pet Store            Beer Store                       Café   
22           Coffee Shop    Chinese Restaurant                   Pharmacy   
40    Turkish Restaurant                   Gym  Latin American Restaurant   
49     Convenience Store    Dim Sum Restaurant   Mediterranean Restaurant   
61           College Gym                  Park                  Bookstore   
63                  Park               Brewery     Furniture / Home Store   
66           Coffee Shop          Tennis Court                 Restaurant   
91         Grocery Store                  Park          Convenience Store   
94   Rental Car Location                Lounge                Coffee Shop   

   5th Most Common Venue   6th Most Common Venue        7th Most Common Venue  \
1                 Lounge                    Park                   Playground   
14       Thai Restaurant             Curling Ice                         Café   
17         Grocery Store  Transportation Service                 Liquor Store   
22     Mobile Phone Shop       Indian Restaurant         Fast Food Restaurant   
40        Sandwich Place            Liquor Store            Other Repair Shop   
49           Gas Station                  Bakery                         Park   
61                  Café                   Trail         Gym / Fitness Center   
63           Gas Station       Convenience Store                     Bus Line   
66     Convenience Store       French Restaurant                         Bank   
91    Athletics & Sports           Metro Station               Breakfast Spot   
94                  Farm     Dumpling Restaurant  Eastern European Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
1            Pizza Place                  Golf Course            Men's Store  
14          Skating Rink               Cosmetics Shop           Liquor Store  
17    College Rec Center                 Intersection      Fish & Chips Shop  
22   Dumpling Restaurant  Eastern European Restaurant      Electronics Store  
40     Electronics Store           Italian Restaurant                   Park  
49    Athletics & Sports           Chinese Restaurant            Pizza Place  
61           Coffee Shop  Eastern European Restaurant      Electronics Store  
63            Beer Store                  Pizza Place      Indian Restaurant  
66               Dog Run                  Golf Course          Grocery Store  
91   Filipino Restaurant                  Candy Store                 Bistro  
94     Electronics Store         Ethiopian Restaurant            Event Space

Cluster 3 covers areas of greatest mixture of coffee shop, parks and fitness areas. Again from the map, these clusters are outside the Toronto City centre but in smaller quantities. Hence the assumption is that cluster 3 is to serve __Special Residential__ areas.

__Cluster 4__

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood  1st Most Common Venue 2nd Most Common Venue  \
53  Downsview-M3M  Vietnamese Restaurant            Food Truck   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
53        Baseball Field  Fast Food Restaurant  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
53     Electronics Store  Ethiopian Restaurant           Event Space   

   9th Most Common Venue 10th Most Common Venue  
53    Falafel Restaurant                   Farm

Cluster 4 covers areas of establishments providing food mostly

__Cluster 5__

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                                        Parkwoods-M3A                  Park   
5                                 Islington Avenue-M9A              Pharmacy   
11   West Deane Park, Princess Gardens, Martin Grov...                  Park   
12          Rouge Hill, Port Union, Highland Creek-M1C        Breakfast Spot   
21                             Caledonia-Fairbanks-M6E              Pharmacy   
27                               Hillcrest Village-M2H           Coffee Shop   
28   Bathurst Manor, Wilson Heights, Downsview Nort...           Pizza Place   
39                                 Bayview Village-M2K   Japanese Restaurant   
46                                       Downsview-M3L                  Park   
50                                   Humber Summit-M9L     Electronics Store   
51   Cliffside, Cliffcrest, Scarborough Village Wes...           Pizza Place   
57                                Humberlea, Emery-M9M                Bakery   
58                     Birch Cliff, Cliffside West-M1N                  Park   
72                                      Willowdale-M2R              Pharmacy   
88    New Toronto, Mimico South, Humber Bay Shores-M8V                  Park   
93                          Alderwood, Long Branch-M8W        Discount Store   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...                  Park   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0           Shopping Mall     Convenience Store                     Pharmacy   
5       Convenience Store                Bakery                Shopping Mall   
11            Pizza Place                 Hotel                   Restaurant   
12           Burger Joint                  Park                   Playground   
21               Bus Stop                  Park                  Pizza Place   
27                   Park              Pharmacy                       Bakery   
28                   Park           Coffee Shop                         Bank   
39          Grocery Store           Gas Station                         Bank   
46                   Bank         Shopping Mall                Moving Target   
50    Arts & Crafts Store              Pharmacy                       Bakery   
51         Ice Cream Shop                 Beach                   Sports Bar   
57      Convenience Store           Golf Course             Storage Facility   
58               Gym Pool                  Café        General Entertainment   
72            Pizza Place                Bakery  Eastern European Restaurant   
88                 Bakery     Indian Restaurant                 Liquor Store   
93               Pharmacy                  Park            Convenience Store   
101    Italian Restaurant         Shopping Mall               Ice Cream Shop   

           5th Most Common Venue  6th Most Common Venue  \
0                       Bus Stop      Fish & Chips Shop   
5                    Golf Course                   Park   
11                Clothing Store          Grocery Store   
12            Italian Restaurant            Fish Market   
21            Mexican Restaurant                   Bank   
27             Korean Restaurant          Shopping Mall   
28                      Pharmacy       Community Center   
39            Chinese Restaurant             Restaurant   
46                   Pizza Place  Vietnamese Restaurant   
50                 Shopping Mall                   Park   
51                  Burger Joint                   Park   
57                          Park            Gas Station   
58             Convenience Store        Thai Restaurant   
72                          Park          Grocery Store   
88                  Skating Rink    Fried Chicken Joint   
93                   Pizza Place            Coffee Shop   
101  Eastern European Restaurant   Gym / Fitness Center   

         7th Most Common Venue        8th Most Common Venue  \
0                  Supermarket    

Cluster 5 covers areas of mostly shopping malls, stores and parks. From the map most are distributed around the outside of Toronto City centre within pockets of Residential area. Cluster 5 can be assumed to be __Retail Parks__.

__Cluster 6__

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
2                        Regent Park, Harbourfront-M5A           Coffee Shop   
3                 Lawrence Manor, Lawrence Heights-M6A        Clothing Store   
4      Queen's Park, Ontario Provincial Government-M7A           Coffee Shop   
7                                        Don Mills-M3B           Coffee Shop   
9                         Garden District, Ryerson-M5B           Coffee Shop   
13                                       Don Mills-M3C            Restaurant   
15                                  St. James Town-M5C           Coffee Shop   
16                              Humewood-Cedarvale-M6C     Convenience Store   
19                                     The Beaches-M4E           Coffee Shop   
20                                     Berczy Park-M5E           Coffee Shop   
23                                         Leaside-M4G           Coffee Shop   
24                              Central Bay Street-M5G           Coffee Shop   
25                                        Christie-M6G     Korean Restaurant   
30                        Richmond, Adelaide, King-M5H           Coffee Shop   
31                    Dufferin, Dovercourt Village-M6H                  Café   
33                    Fairview, Henry Farm, Oriole-M2J        Clothing Store   
35                                    East Toronto-M4J           Coffee Shop   
36   Harbourfront East, Union Station, Toronto Isla...           Coffee Shop   
37                        Little Portugal, Trinity-M6J                  Café   
41                    The Danforth West, Riverdale-M4K      Greek Restaurant   
42        Toronto Dominion Centre, Design Exchange-M5K           Coffee Shop   
43    Brockton, Parkdale Village, Exhibition Place-M6K                  Café   
47                  India Bazaar, The Beaches West-M4L     Indian Restaurant   
48                  Commerce Court, Victoria Hotel-M5L           Coffee Shop   
52                         Willowdale, Newtonbrook-M2M     Korean Restaurant   
54                                 Studio District-M4M           Coffee Shop   
55               Bedford Park, Lawrence Manor East-M5M    Italian Restaurant   
59                                      Willowdale-M2N       Bubble Tea Shop   
62                                        Roselawn-M5N      Sushi Restaurant   
67                                Davisville North-M4P           Coffee Shop   
68                        Forest Hill North & West-M5P                  Park   
69                   High Park, The Junction South-M6P                  Café   
73                              North Toronto West-M4R    Italian Restaurant   
74             The Annex, North Midtown, Yorkville-M5R                  Café   
75                          Parkdale, Roncesvalles-M6R                  Café   
76           Canada Post Gateway Processing Centre-M7R           Coffee Shop   
79                                      Davisville-M4S    Italian Restaurant   
80                  University of Toronto, Harbord-M5S                  Café   
81                              Runnymede, Swansea-M6S           Coffee Shop   
83                     Moore Park, Summerhill East-M4T           Coffee Shop   
84       Kensington Market, Chinatown, Grange Park-M5T                  Café   
86   Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
87   CN Tower, King and Spadina, Railway Lands, Har...                  Café   
92                                  Stn A PO Boxes-M5W           Coffee Shop   
96                     St. James Town, Cabbagetown-M4X                  Park   
97          First Canadian Place, Underground city-M5X           Coffee Shop   
98   The Kingsway, Montgomery Road, Old Mill North-M8X           Coffee Shop   
99                            Church and Wellesley-M4Y           Coffee Shop   
100          Business reply mail Processing Centre-M7Y                  Park   
102  Mimico NW, The Queensway West, South 

Cluster 6 covers areas of greatest coffee shops, cafes and restaurants. Looking at the map, the cluster are generally centred around the Toronto City centre. Hence, the assumption is that this cluster is mainly __Downtown and Business__ areas